src/core/public/application/application_service.test.tsx

```diff
@@ -25,211 +25,491 @@ import { MockCapabilitiesService, MockHistory } from './application_service.test
+          Map {
+            "app1" => Object {
+              "id": "app1",
+              "legacy": false,
+              "status": 1,
+              "tooltip": "App inaccessible due to reason",
+            },
+            "app2" => Object {
+              "id": "app2",
+              "legacy": false,
+              "status": 0,
+            },
+          }
+        `);
+      });
     });
 
-    it('throws error if additional apps are registered after setup', async () => {
-      const service = new ApplicationService();
-      const context = contextServiceMock.createSetupContract();
-      const setup = service.setup({ context });
-      const http = httpServiceMock.createStartContract();
-      const injectedMetadata = injectedMetadataServiceMock.createStartContract();
-      await service.start({ http, injectedMetadata });
-      expect(() =>
-        setup.registerLegacyApp({ id: 'app2' } as any)
-      ).toThrowErrorMatchingInlineSnapshot(`"Applications cannot be registered after \\"setup\\""`);
+    describe('registerLegacyApp', () => {
+      it('throws an error if two apps with the same id are registered', () => {
+        const setup = service.setup({ context });
+        setup.registerLegacyApp({ id: 'app2' } as any);
+        expect(() =>
+          setup.registerLegacyApp({ id: 'app2' } as any)
+        ).toThrowErrorMatchingInlineSnapshot(
+          `"A legacy application is already registered with the id \\"app2\\""`
+        );
+      });
+
+      it('throws error if additional apps are registered after setup', async () => {
+        const setup = service.setup({ context });
+        await service.start({ http, injectedMetadata });
+        expect(() =>
+          setup.registerLegacyApp({ id: 'app2' } as any)
+        ).toThrowErrorMatchingInlineSnapshot(
+          `"Applications cannot be registered after \\"setup\\""`
+        );
+      });
     });
-  });
 
-  it("`registerMountContext` calls context container's registerContext", () => {
-    const service = new ApplicationService();
-    const context = contextServiceMock.createSetupContract();
-    const setup = service.setup({ context });
-    const container = context.createContextContainer.mock.results[0].value;
-    const pluginId = Symbol();
-    const noop = () => {};
-    setup.registerMountContext(pluginId, 'test' as any, noop as any);
-    expect(container.registerContext).toHaveBeenCalledWith(pluginId, 'test', noop);
-  });
-});
+    describe('registerAppStatusUpdater', () => {
+      it('updates status fields', async () => {
+        const setup = service.setup({ context });
 
-describe('#start()', () => {
-  beforeEach(() => {
-    MockHistory.push.mockReset();
-  });
+        const pluginId = Symbol('plugin');
+        setup.register(pluginId, { id: 'app1' } as any);
+        setup.register(pluginId, { id: 'app2' } as any);
+        setup.registerAppUpdater(
+          new BehaviorSubject<AppUpdater>(app => {
+            if (app.id === 'app1') {
```

Comment:
nit: could you add a second branch `if (app.id === 'app1') {` to make sure it's called for **every** app ?